### PR TITLE
Fix the addxml  when add node after comment xml node

### PR DESCRIPTION
**Pull request by JaynieBai**
Fixes [#11933 ](https://github.com/dotnet/msbuild/issues/11933)

### Context
```
<ItemGroup>
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" /><!-- some comment -->
  </ItemGroup>
```
**dotnet add package Serilog --version 4.3.0
The trailing comment on the Newtonsoft.Json line is moved to the newly added package.**

```
<ItemGroup>
  <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
  <PackageReference Include="Serilog" Version="4.3.0" /><!-- some comment -->
</ItemGroup>
```

### Changes Made
If the reference element is immediately followed by a series of XML nodes, each of which does not contain any line breaks and is whitespace, a comment, or a processing instruction, then insert the new element after those nodes.

### Testing
AddPackageReference_PreservesTrailingCommentOnSameLine

### CODE REVIEW

File: src/Build/Construction/ProjectElementContainer.cs [hunk 1]
```diff
@@ -521,16 +521,29 @@ internal void AddToXml(ProjectElement child)
...
-                    XmlElement.InsertAfter(child.XmlElement, referenceSibling.XmlElement);
+                    // Heuristic: If the reference element is immediately followed by a series of XML nodes, each of which
+                    // does not contain any line breaks and is whitespace, a comment, or a processing instruction,
+                    // then insert the new element after those nodes.
+                    XmlNode insertAfter = referenceSibling.XmlElement;
+                    XmlNode next = insertAfter.NextSibling;
+                    while (next != null &&
+                        (next.NodeType == XmlNodeType.Whitespace ||
+                         next.NodeType == XmlNodeType.Comment ||
+                         next.NodeType == XmlNodeType.ProcessingInstruction)
```
Comment (AR-May) [inline src/Build/Construction/ProjectElementContainer.cs]: Why do we need `XmlNodeType.ProcessingInstruction` in the condition? I have never saw them in the middle of the project file, is there some examples when this would be needed?

File: src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs [hunk 2]
```diff
@@ -102,5 +104,36 @@ public void SetLabel()
...
         }
+
+        [Fact]
+        public void AddPackageReference_PreservesTrailingCommentOnSameLine()
+        {
+            string content =
+            """
+               <Project>
+                   <ItemGroup>
+                       <PackageReference Include="Newtonsoft.Json" Version="13.0.1" /><!-- some comment -->
```
Comment (AR-May) [inline src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs]: Could you please add more tests, involving comments? Add please case of multi-line comments before the next item. Add also second item, so that you insert between them.

**Review by AR-May — CHANGES_REQUESTED** [review]
This change feels like a workaround, which makes me wonder if we might be overlooking some edge cases. It would be helpful if you could share more examples of how the code behaves, either by adding test cases or simply including them as comments.

File: src/Build/Construction/ProjectElementContainer.cs (hunk 1 above)
Comment (JaynieBai) [reply src/Build/Construction/ProjectElementContainer.cs]: Yes, you are right, XmlNodeType.ProcessingInstruction is not supported in the project file. So, I remove that.

File: src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs (hunk 2 above)
Comment (JaynieBai) [reply src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs]: Add 3 more scenarios in the test AddItem_PreservesComments_VariousCases()

File: src/Build/Construction/ProjectElementContainer.cs [hunk 3]
```diff
@@ -521,16 +521,29 @@ internal void AddToXml(ProjectElement child)
...
+                         && (next.Value == null || (!next.Value.Contains("\n") && !next.Value.Contains("\r"))))
```
[the remark this answers is not in the document]
Comment (JaynieBai) [reply src/Build/Construction/ProjectElementContainer.cs]: It's not necessary.

File: src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs [hunk 4]
```diff
@@ -102,5 +104,136 @@ public void SetLabel()
...
+            var newItem = project.CreateItemElement("PackageReference");
+            newItem.Include = "Inserted";
+            group.InsertAfterChild(newItem, items[0]);
+            newItem.AddMetadata("Version", "1.5.0", true);
+
+            Helpers.VerifyAssertLineByLine(expectedContent, project.RawXml);
+        }
+
+        [Fact]
+        public void AddItem_PreservesComments_VariousCases()
```
Comment (AR-May) [inline src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs]: Could you please consolidate both of the tests, so that you do not repeat the code? You can check various cases using `Theory` instead of the `Fact` and passing different content to the test, like [here](https://github.com/dotnet/msbuild/blob/17babbf859e410fb09eaec1e84a0c6e375368641/src/Build.OM.UnitTests/Construction/ProjectItemElement_Tests.cs#L158).

Comment (JaynieBai) [reply src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs]: Updated

File: src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs [hunk 5]
```diff
@@ -102,5 +104,120 @@ public void SetLabel()
             Assert.Equal("c", itemGroup.Label);
             Assert.True(project.HasUnsavedChanges);
         }
+
+        [Fact]
+        public void AddPackageReference_PreservesTrailingCommentOnSameLine()
```
Comment (AR-May) [inline src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs]: You can add this test to the theory as well and I would suggest to rename it `AddPackageReference_PreservesTrailingComments`

Comment (JaynieBai) [reply src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs]: It calls group.AddItem("PackageReference", "Serilog").AddMetadata("Version", "4.3.0", expressAsAttribute: true); in the test method. In the theory, it calls InsertAfterChild directly. So, I separated them into different methods.  But we can combine them together. Have updated

Comment (AR-May) [reply src/Build.OM.UnitTests/Construction/ProjectItemGroupElement_tests.cs]: I guess for testing our behavior it does not matter whether we call it directly or not.

**Review by rainersigwald — APPROVED** [review]

**Review by JanProvaznik — APPROVED** [review]
from an llm:
> The change correctly handles inserting after an element with a trailing comment. However, it misses the symmetric case: inserting before an element that has a leading comment. This is the most significant improvement that should be made.
> Consider this XML:
```
<ItemGroup>
  <Item Include="a.cs" />
  <!-- This is a comment for c.cs -->
  <Item Include="c.cs" />
</ItemGroup>

```

If we call InsertBeforeChild(b_item, c_item), the current PR code (and the main branch) would search right to find c.cs as the reference sibling and then call XmlElement.InsertBefore(child.XmlElement, referenceSibling.XmlElement).
This would produce the following incorrect output:
```
<ItemGroup>
  <Item Include="a.cs" />
  <!-- This is a comment for c.cs -->
  <Item Include="b.cs" /> 
  <Item Include="c.cs" />
</ItemGroup>
```

though I guess that symmetric case does not matter, since the motivation for the PR is addressed by the change
